### DISCUSSION TITLE
How to implement reusable part of form

### DISCUSSION THREAD

**Discussion by nucleartux** [react-hook-form/react-hook-form#4672] [Q&A] [answered]
https://codesandbox.io/s/react-hook-form-control-forked-d8651

Now I'm getting error `Type 'Control<FormValues>' is not assignable to type 'Control<FormPart>'.`

How to fix types? What is the best approach to reuse some elements (including arrays) in different forms?

**Comment by nucleartux** [comment]
@bluebill1049 any thoughts?

**Reply by jorisre** [reply]
Hi @nucleartux 

You have to replace the `FormPart` by `FormValues` because you might use `useController` or `Controller` and specify yourself the input `name`.

```tsx
function ReusableFormPart({ control }: { control: Control<FormValues> }) {
  return <div>todo</div>;
}
```

As you can see, the `name` property is strongly typed.
![Screenshot 2021-04-07 at 17 20 35](https://user-images.githubusercontent.com/7545547/113891678-93dcba80-97c5-11eb-87ba-b978741f6f90.png)

Does it help you ?

**Reply by nucleartux** [reply]
Sorry for misunderstanding. I don't want to implement only one field, I want to embed whole part of form (multiple fields).
Also, they might be in different parts: for example I have FormPart on top level, and also want to use it inside field array of same form.

**Reply by jorisre** [reply]
Ok I know what you mean, unfortunately I don't know how to do that, I tried with generic types without success.
@bluebill1049 it's your turn :)

**Reply by bluebill1049** [reply]
@nucleartux you can still use FormValues as the generic right? it's not required to have all the input's name, I was wondering why you need to have individual types against your component.

**Reply by nucleartux** [reply]
Look:
I have Form1 with FormValues1 (for example with name, surname, city).
I have Form2 with FormValues2 (for example with name, surname, birthday).
I want to reuse part with name and surname. How can I achieve this?

**Reply by cabbiepete** [reply]
FWIW I found the simplest solution was to pass the FormFields type definition around. 

This codesandbox shows how to resolve using this technique. https://codesandbox.io/s/mystifying-lichterman-w1h45?file=/src/ControlledTextField.tsx

In your case that would mean Form1 and Form2 would both be generic types with TFieldValues as a type parameter.

When you use Form1/Form2 you'll call it like `<Form1<FormFields> ... />`, `<Form2<FormFields> ... />` where FormFields is something like 

```
type FormFields = {
  name: string
  surname: string
  city: string
  birthday: string
}
```

**Reply by IanVS** [reply]
> @nucleartux you can still use FormValues as the generic right? it's not required to have all the input's name, I was wondering why you need to have individual types against your component.

@bluebill1049 as far as I can tell, the `FormValues` has to be exact, not a subset or superset.  See https://codesandbox.io/s/react-hook-form-context-3qjt9j?file=/src/App.tsx.  I've tried it three different ways, and each way has a typescript error.  Would _love_ to find the right way to do this.

**Comment by cabbiepete** [comment, marked as answer]
I've updated the original codesandbox with with I think is about as good as we can do any improvements anyone can see I'd love to see them :-)

https://codesandbox.io/s/react-hook-form-control-forked-gfm98

**Reply by IanVS** [reply]
Unfortunately that doesn't seem to work.  I want to be able to `register` inputs in my sub-form, and while it works fine in js, the TS types are not happy.  I've tweaked the sandbox above to demonstrate: https://codesandbox.io/s/react-hook-form-context-3qjt9j

```
Argument of type '"firstName"' is not assignable to parameter of type 'Path<TFieldValues>'
```

**Reply by IanVS** [reply]
Oh, it seems like this is a known issue: https://github.com/react-hook-form/react-hook-form/issues/6726.  Bummer, seems like this kind of reusable form partial is not possible?
